tests/python/pants_test/engine/test_scheduler.py

```diff
@@ -37,14 +37,37 @@ def consumes_a_and_b(a, b):
   return str('{} and {}'.format(a, b))
 
 
+class C(object):
+  pass
+
+
+@rule(B, [Select(C)])
+def transitive_b_c(c):
+  return B()
+
+
+class D(object):
+  pass
+
+
+@rule(D, [Select(C)])
+def transitive_coroutine_rule(c):
+  yield Get(B, C, c)
```

Comment:
hm......given that this is all `ast` module magic in the first place it might make sense to require `return` statements in the final position here. I will look into implementing this under the assumption it will be quick enough. After reading your comment I also am now perplexed by the assignment-less `yield` at the end.

Reply:
Thinking further (especially after seeing that "assignment-less" yields are good for (admittedly, rather hasty) rules in #6891), it might be appropriate to disallow assignment-less yields entirely and force the use of e.g. `_ = yield Get(...)` instead (this would still disallow a final assignment-less `yield` statement as well).

Reply:
A function isn't allowed to both yield *and* return (I think this may be changing in 3.8?), which is why we have these awkwardness. I would be in favour of requiring that a LHS-less yield must either be the last statement in the function, or followed by a literal `return` statement at the same level of nesting, though, which we could enforce in our AST magic...

(This is one place where not being constrained by python syntax would be really nice)

Reply:
> it might be appropriate to disallow assignment-less yields entirely and force the use of e.g. _ = yield Get(...) instead (this would still disallow a final assignment-less yield statement as well).

This might be reasonable, but I think that it would add a fair amount of complexity to our parsing. We don't actually currently associate `Gets` with `yield` statements: instead, we only parse the constructor invoke itself, regardless of location in the method body.

Reply:
But as to convention: it would be cleaner to write `_ = yield Get(..)` in general.

Reply:
I'm looking at the ast magic currently -- my interpretation was similar to yours @illicitonion, except for:

> followed by a literal return statement at the same level of nesting

Are you referring to something like:

```python
@rule(...)
def f(...):
  # ...
  yield Get(...)
  return 3
```

Would this merely be a convenience over:

```python
@rule(...)
def f(...):
  # ...
  _ = yield Get(...)
  return 3
```

Or am I missing something? I'm currently predisposed to feel like allowing a bare `yield` at all is inconsistent in some sense, but not too opinionated.

Reply:
> This might be reasonable, but I think that it would add a fair amount of complexity to our parsing.

@stuhood (missed your comment earlier, sorry): I just happened to find the `ast.NodeTransformer` API not too difficult to use and was looking into that. Since the `yield` would be a single level lower than the assignment, I'm thinking the checking wouldn't be super hard, maybe. It is entirely up in the air at this point and implementation complexity/brittleness should and will be weighed against the utility of clearly describing a convention (as you have here).

> It is not legal python to write ...

@illicitonion I hadn't realized this!

```python
yield 3
return
```

The description you provided for the above construction makes complete sense to me here.

Reply:
I would not be in favor of re-writing anything, but I might be in favor of banning certain patterns (ie, a `yield Get` without an LHS). IMO, we should not be diverging from valid python syntax in any way: we should ensure that we are a subset.

Reply:
> we should not be diverging from valid python syntax in any way: we should ensure that we are a subset.

I strongly vibe with this approach and will bound this to banning `yield Get` without an LHS (would we leave `yield`s that aren't for `Get`s unchanged?). I'm timeboxing this to the next hour.

Reply:
It took 2 hours to make really solid but 1dde22f implements the checking described here and edits `@rule` bodies throughout the repo to have an empty `return` after any closing `yield`.

Reply:
> and edits @rule bodies throughout the repo to have an empty return after any closing yield.

I'm sorry, but I'm not in favor of that at all. I see now that Daniel and I were suggesting different things, but the stray `return`s do not improve readability, IMO.

I was suggesting banning:
```
# This will result in a discarded value, because yielding a Get causes a value to be produced.
yield Get(...)
```
...by requiring _some_ LHS:
```
# Requiring an LHS on yields that contain `Get`s makes it obvious that a value is being discarded.
_ = yield Get(...)
```

But as mentioned, that is hard.

Sorry for the miscommunication. I'd suggest splitting out the syntax changes to another PR for independent discussion.